xpublish_edr/select.py

```diff
@@ -0,0 +1,125 @@
+            elif len(datetimes) == 2:
+                ds = ds.cf.sel(T=slice(datetimes[0], datetimes[1]))
+            else:
+                raise ValueError(
+                    f"Invalid datetimes submitted - {datetimes}",
+                )
+        except ValueError as e:
+            logger.error("Error with datetime", exc_info=True)
+            raise ValueError(f"Invalid datetime ({e})") from e
+
+    if query.parameters:
+        try:
+            ds = ds.cf[query.parameters.split(",")]
+        except KeyError as e:
+            raise ValueError(f"Invalid variable: {e}") from e
+
+    query_param_keys = list(query_params.keys())
+    for query_param in query_param_keys:
+        if query_param in edr_query_params:
+            del query_params[query_param]
+
+    method: Optional[str] = "nearest"
+
+    for key, value in query_params.items():
+        split_value = value.split("/")
+        if len(split_value) == 1:
+            continue
+        elif len(split_value) == 2:
+            query_params[key] = slice(split_value[0], split_value[1])
+            method = None
+        else:
+            raise ValueError(f"Too many values for selecting {key}")
+
+    ds = ds.sel(query_params, method=method)
+    return ds
+
+
+def select_postition(ds: xr.Dataset, point: Point) -> xr.Dataset:
+    """
+    Return a dataset with the position nearest to the given coordinates
+    """
+    if _is_regular_xy_coords(ds):
+        return _select_position_regular_xy_grid(ds, point)
+    else:
+        # TODO: Handle 2D coordinates
+        raise NotImplementedError("Only 1D coordinates are supported")
+
+
+def select_area(ds: xr.Dataset, polygon: Polygon) -> xr.Dataset:
+    """
+    Return a dataset with the area within the given polygon
+    """
+    if _is_regular_xy_coords(ds):
+        return _select_area_regular_xy_grid(ds, polygon)
+    else:
+        # TODO: Handle 2D coordinates
+        raise NotImplementedError("Only 1D coordinates are supported")
+
+
+def _coord_is_regular(da: xr.DataArray) -> bool:
+    """
+    Check if the DataArray has a regular grid
+    """
+    return len(da.shape) == 1 and da.name in da.dims
+
+
+def _is_regular_xy_coords(ds: xr.Dataset) -> bool:
+    """
+    Check if the dataset has 2D coordinates
+    """
+    return _coord_is_regular(ds.cf["X"]) and _coord_is_regular(ds.cf["Y"])
+
+
+def _select_position_regular_xy_grid(ds: xr.Dataset, point: Point) -> xr.Dataset:
+    """
+    Return a dataset with the position nearest to the given coordinates
+    """
+    # Find the nearest X and Y coordinates to the point
+    return ds.cf.sel(X=point.x, Y=point.y, method="nearest")
+
+
+def _select_area_regular_xy_grid(ds: xr.Dataset, polygon: Polygon) -> xr.Dataset:
+    """
+    Return a dataset with the area within the given polygon
+    """
+    # For a regular grid, we can create a meshgrid of the X and Y coordinates to create a spatial mask
+    pts = np.meshgrid(ds.cf["X"], ds.cf["Y"])
```

Comment:
I would clip these to `polygon.bounds` to minimize the perf hit of `np.vectorize`

https://shapely.readthedocs.io/en/stable/reference/shapely.Polygon.html#shapely.Polygon.bounds

Reply:
great idea, the downselection happens later anyways so might as well do it up front